ext/pdo_firebird/firebird_driver.c

```diff
@@ -460,16 +460,67 @@ int preprocess(const zend_string* sql, char* sql_out, HashTable* named_params)
 }
 
 /* map driver specific error message to PDO error */
-void _firebird_error(pdo_dbh_t *dbh, pdo_stmt_t *stmt, char const *file, zend_long line) /* {{{ */
+void _firebird_error(pdo_dbh_t *dbh, pdo_stmt_t *stmt, const char *state, const size_t state_len,
+	const char *msg, const size_t msg_len) /* {{{ */
 {
 	pdo_error_type *const error_code = stmt ? &stmt->error_code : &dbh->error_code;
+	pdo_firebird_db_handle *H = (pdo_firebird_db_handle *)dbh->driver_data;
+	pdo_firebird_error_info *einfo = &H->einfo;
+	int sqlcode = -999;
+
+	if (einfo->errmsg) {
+		pefree(einfo->errmsg, dbh->is_persistent);
+		einfo->errmsg = NULL;
+		einfo->errmsg_length = 0;
+	}
+
+	if (H->isc_status && (H->isc_status[0] == 1 && H->isc_status[1] > 0)) {
+		char buf[512];
+		size_t buf_size = sizeof(buf), read_len = 0;
+		ssize_t tmp_len;
+		const ISC_STATUS *s = H->isc_status;
+		sqlcode = isc_sqlcode(s);
+
+		while ((buf_size > (read_len + 1)) && (tmp_len = fb_interpret(&buf[read_len], (buf_size - read_len - 1), &s)) && tmp_len > 0) {
+			read_len += tmp_len;
+			buf[read_len++] = ' ';
+		}
 
-	strcpy(*error_code, "HY000");
+		/* remove last space */
+		if (read_len) {
+			buf[read_len--] = '\0';
+		}
+
+		einfo->errmsg_length = read_len;
+		einfo->errmsg = pestrndup(buf, read_len, dbh->is_persistent);
+
+#if FB_API_VER >= 25
+		char sqlstate[sizeof(pdo_error_type)];
+		fb_sqlstate(sqlstate, H->isc_status);
+		if (sqlstate != NULL && strlen(sqlstate) < sizeof(pdo_error_type)) {
+			strcpy(*error_code, sqlstate);
+			goto end;
+		}
+#endif
+	} else if (msg && msg_len) {
+		einfo->errmsg_length = msg_len;
+		einfo->errmsg = pestrndup(msg, einfo->errmsg_length, dbh->is_persistent);
+	}
+
+	if (state && state_len && state_len < sizeof(pdo_error_type)) {
+		memcpy(*error_code, state, state_len + 1);
+	} else {
+		memcpy(*error_code, "HY000", 6);
```

Comment:
This *should* be better AFAIK
```suggestion
		memcpy(*error_code, "HY000", sizeof("HY000");
```